Fix setup and add new schema

JosePizarro3 opened the pull request:
@jkshenton Sathya (sorry, I cannot tag him as he is not part of this repo),

I've been working on some changes in this repo. Maybe you can drop me a review so I can merge and add this new plugin into NOMAD? You can also add it then as part of the NOMAD OASIS and use it in your CCP-NC database.

I still need to:
- [x] Add testing for the parser under `tests/test_parser.py`
- [x] Fix workflow recognition with CASTEP output files

### Changes 1 - new structure
I've adopted the new structure for plugins from NOMAD. This means a lot of changes, but the most important are the ones in `pyproject.toml` and the new folder structure under `src`. Some extra stuff:

- You can release this project in PyPI and maintain it with versioning in the way you want. 
- I might need to do minor changes from time to time to reflect changes in `nomad`.
- I added instructions in the README.md, so you can test if these work or not for you.
- I added a MkDocs page with a NOMAD style, that you can of course delete or restructure at will.

### Changes 2 - New schema `nomad-simulations`
I've changed the parsing to the new schema, [`nomad-simulations`](https://github.com/nomad-coe/nomad-simulations). Some points here:

- You are welcome to adopt or not this schema, as we already talked. If you need more quantities or a different way of parsing, I suggest you work from something up. The example here would be how I used `Outputs` from the `nomad-simulations` schema to **extend it** with NMR specific properties, and how I used `PhysicalProperty` from this package to define such NMR properties.
- Now, I think the properties are much nicer parsed, as they contained references to the specific atomic states, more similar to what you have in the magres file.
- In FAIRmat we are developing right now a taxonomy for materials properties, so anyways these properties might end up defined in the `nomad` or `nomad-simulations` repo, as some other users might be using these anyways.
- As you can see, these properties have some functionalities to extract information (like the isotropic part of the MagneticShielding tensor), and these could be extended. I left it for you to decide whether you want to do this here or in some other repo you have.
- Ideally, and to keep a nice collaboration, these properties could be perfectly defined in the `nomad-simulations` repo. I will be happy if you want to contribute there and _update_ these definitions in there. Other users might need them and it connects nicely with the taxonomy effort I comment on above.

JosePizarro3 committed:
Initial structuring

JosePizarro3 committed:
Added nmr output properties into the package definitions

JosePizarro3 committed:
Deleted legacy runschema

JosePizarro3 committed:
Added parsing into the new nomad-simulations schema

JosePizarro3 committed:
Added docstrings

JosePizarro3 committed:
Adopting child_archives matching interface

JosePizarro3 committed:
Final touches on package.py

JosePizarro3 committed:
Fixed github ci/cd

jkshenton commented:
Thanks a lot @JosePizarro3 - really appreciate your continued help with this! The restructuring looks fine to me. The new schema (atoms state thing) will be much easier for us to work with I suspect and in general makes a lot of sense. 

In terms of contributing to nomad-simulations, we are absolutely keen. The [magres format spec document](https://www.ccpnc.ac.uk/docs/magres/magres-format.pdf) was already some effort to defining a standard schema for computed magres data for the community. I would be very happy to see it adapted for nomad generally. I count >13 periodic electronic structure codes that can calculate magnetic shielding tensors (one of the key properties) so having a general schema that other parsers can work with makes sense to me. 

I've manually added @Sathya-S3  now and I think he's in a better position to review the specific code changes having tested out the initial version.

JosePizarro3 commented:
Great, thanks @jkshenton !

> In terms of contributing to nomad-simulations, we are absolutely keen. The [magres format spec document](https://www.ccpnc.ac.uk/docs/magres/magres-format.pdf) was already some effort to defining a standard schema for computed magres data for the community. I would be very happy to see it adapted for nomad generally. I count >13 periodic electronic structure codes that can calculate magnetic shielding tensors (one of the key properties) so having a general schema that other parsers can work with makes sense to me.

Indeed, the magres format is a very easy case where, whatever you decided for the properties should be incorporated directly, while for the symmetry open questions, we should work a bit more on it.

JosePizarro3 committed:
Added tests

JosePizarro3 committed:
Fix mypy

JosePizarro3 committed:
Added automatic connection with CASTEP entries

Added nomad-lab[infrastructure] extra dependencies

Added workflow for NMR CASTEP or QE + magres entry

JosePizarro3 committed:
Added docstring to init_parser

JosePizarro3 committed:
Fix import circular problem

Sathya-S3 commented on pyproject.toml:
```diff
@@ -1,63 +1,95 @@
-classifiers = [
-    "Programming Language :: Python :: 3",
-    "Programming Language :: Python :: 3 :: Only",
-    "Programming Language :: Python :: 3.9",
-    "License :: OSI Approved :: Apache Software License",
+maintainers = [
+    { name = "The CCP-NC Authors" },
```
Small query: Is the intended result for users to contact the CCP-NC authors directly via GitHub or is this text a placeholder to add names and emails later?

JosePizarro3 replied:
So it seems that this does not for the PyPI page. I thought it would show these and the authors as a list, but we can keep it as it is now due to its little impact

Sathya-S3 commented on pyproject.toml:
```diff
@@ -1,63 +1,95 @@
-"Documentation" = "https://nomad-coe.github.io/nomad-parser-magres/"
+"Homepage" = "https://github.com/CCP-NC/nomad-parser-magres"
+"Bug Tracker" = "https://github.com/CCP-NC/nomad-parser-magres/issues"
+"Documentation" = "https://CCP-NC.github.io/nomad-parser-magres/"
```
The documentation page "https://CCP-NC.github.io/nomad-parser-magres/" does not exist at the time of review. I assume this will be auto-rectified when the changes are merged into "master branch". Will run a quick verification after merge.

JosePizarro3 replied:
we solved this last time we talk, the docu page has to be deployed in the settings of the repo. Check this: https://www.mkdocs.org/user-guide/deploying-your-docs/#github-pages

Sathya-S3 replied:
@jkshenton for your attention

Sathya-S3 commented on src/nomad_parser_magres/parsers/parser.py:
```diff
@@ -0,0 +1,740 @@
+        workflow.name = 'NMR Magres'
+
+        # ! Fix this once CASTEP and QuantumESPRESSO use the new `nomad-simulations` schema under 'data'
+        # Method
+        # method_nmr = extract_section(nmr_first_principles_archive, ['run', 'method'])
+        # workflow.method.nmr_method_ref = method_nmr
+
+        # Inputs and Outputs
+        # ! Fix this to extract `input_structure` from `nmr_first_principles_archive` once
+        # ! CASTEP and QuantumESPRESSO use the new `nomad-simulations` schema under 'data'
+        input_structure = extract_section(self.archive, ['data', 'model_system'])
+        nmr_magres_calculation = extract_section(self.archive, ['data', 'outputs'])
+        if input_structure:
+            workflow.m_add_sub_section(
+                NMRMagRes.inputs, Link(name='Input structure', section=input_structure)
+            )
+        if nmr_magres_calculation:
+            workflow.m_add_sub_section(
+                NMRMagRes.outputs,
+                Link(name='Output NMR calculation', section=nmr_magres_calculation),
+            )
+
+        # NMR (first principles) task
+        # ! Fix this once CASTEP and QuantumESPRESSO use the new `nomad-simulations` schema under 'data'
+        program_name = nmr_first_principles_archive.run[-1].program.name
+        if nmr_first_principles_archive.workflow2:
+            task = TaskReference(task=nmr_first_principles_archive.workflow2)
+            task.name = f'NMR FirstPrinciples {program_name}'
+            if input_structure:
+                task.inputs = [Link(name='Input structure', section=input_structure)]
+            if nmr_magres_calculation:
+                task.outputs = [
+                    Link(
+                        name='Output NMR calculation',
+                        section=nmr_magres_calculation,
+                    )
+                ]
+            workflow.m_add_sub_section(NMRMagRes.tasks, task)
+
+        self.archive.workflow2 = workflow
```
At the moment, the workflow separates the .castep (NMR first principles calculation) file and .magres file resulting from the calculation as two individual entries, with both having display names "<formula> CASTEP NMR SinglePoint Simulation". Ideally, since the magres file is a direct output of the CASTEP calculation, it might be beneficial to include it directly to the outputs of the CASTEP first principles calculation in the workflow.

Will the "fix" mentioned in the code comments correspond to fixing how the workflow is processed currently? Or is it related only to the adoption of the new 'nomad-simulations' schema?

JosePizarro3 replied:
Ok, after we talked: we agreed that the separation is somewhat an artifact, but we can keep it just in case someone uploads simply a magres file. In any case, I have to fix this workflow part; it is not working right now but I know how to handle it.

Sathya-S3 commented on src/nomad_parser_magres/parsers/parser.py:
```diff
@@ -0,0 +1,740 @@
+        if outputs is not None:
+            simulation.outputs.append(outputs)
+
+        # Try to resolve the `entry_id` and `mainfile` of other entries in the upload to connect the magres entry with the CASTEP or QuantumESPRESSO entry
```
In the metadata section of file upload, there is a files section that lists all files included in the upload. Is this the only intended result? There is mention of linking the NMR workflow directly in the magres entry, could you please clarify how it is done? As per my previous comments, we see two separate file entries and workflows for the .castep and .magres files.

JosePizarro3 replied:
These lines are more to connect the castep with the magres entries

Sathya-S3 commented on src/nomad_parser_magres/schema_packages/workflow.py:
```diff
@@ -1,27 +1,9 @@
-#
-from nomad.metainfo import SubSection, Quantity, Reference
-from nomad.datamodel.metainfo.simulation.method import Method
+from nomad.metainfo import Quantity, Reference, SubSection
```
This NMRMagRes class is defined to generate an extra EntryArchive if both NMR first principles and NMR magres SinglePoint EntryArchives are present. Could you please clarify how this works in linking the workflows as mentioned in the parser.py script?

JosePizarro3 replied:
The idea's that the magres file creates an entry which connects with the CASTEP or QE one. In this entry, you will find all the details of the original DFT calculation + the output nmr properties.

The creation of such a workflow in the magres entry is just an organizational tool. This is actually what workflows should be about, just organizing in one way without cluttering entries which could be reuse for something else (in this case, the CASTEP or QE ones).

In any case, this can be changed later. We could decide to integrate the workflow better, hence just adding a few lines in the CASTEP and QE parsers to handle the possibility of a magres output file.

Sathya-S3 reviewed (commented):
Thank you for your work on presenting the magres parser in such a polished state, Jose. My main comments are related to understanding the workflow links between CASTEP and magres files, which I have highlighted with the relevant code. I'm happy to approve the pull request after we have had a chance to chat with you about the workflow linking and how we can help with further magres parser development (Quantum Espresso workflows etc.).

I have integrated the magres parser as guided by the README instructions with a NOMAD Oasis instance and have verified that the file parsing works flawlessly. The test was performed with an ethanol CASTEP-magres workflow files archive.

Sathya-S3 commented on README.md:
````diff
@@ -1,128 +1,159 @@
+If you want to develop locally this package, clone the project and in the workspace folder, create a virtual environment (note this project uses Python 3.9):
 ```sh
-git clone https://github.com/nomad-coe/nomad-parser-magres.git
+git clone https://github.com/CCP-NC/nomad-parser-magres.git
````
Should this url point to the CCP-NC GitHub page?

JosePizarro3 replied:
The CCP-NC one, where the repo is hosted right now.

JosePizarro3 commented:
@Sathya-S3 I will do a final clean up next week and push again. I think I can integrate your comments.

JosePizarro3 commented:
I totally forgot about this PR. @Sathya-S3 reminded me and asked me a couple of things, so the problem is that the workflow might have some issues. I can follow https://github.com/FAIRmat-NFDI/nomad-parser-wannier90/pull/3 to fix it and this will be ready.

Sathya-S3 committed:
Adding custom metadata

JosePizarro3 committed:
Fixed ccpnc_metadata schema

Added new CCPNCSimulation class

Fix dependencies

Added get_files utility function

Fix README

JosePizarro3 committed:
Fix ruff

JosePizarro3 commented:
@Sathya-S3 @jkshenton 

Ok, I finished with some changes. Note the changes in `src/schema_packages/ccpnc_metada.py`, `src/schema_packages/package.py` and the last lines in `src/parsers/parser.py`. What I did:

- In ccpnc_metadata.py: fix some types (instead of tuples use `JSON`) and shapes (if the shape is empty, no need to define). Add 2 comments.
- In package.py: add `CCPNCSimulation` at the end.
- In parser.py: change import to use `CCPNCSimulation` instead of `nomad_simulations.Simulation`. Also, I added some lines at the end of the file to clarify how the parsing of the extra CCP-NC metadata could look like; if you decide to go with another option, that's fine for me as well :-)

If you both agree, we can merge this.

jkshenton commented:
Thanks so much again for your help with this, @JosePizarro3 ! I am happy to merge now and make any further changes in subsequent PRs.
